Shelfice controls, generic masking, and lognormal control

timothyas opened the pull request:
## What changes does this PR introduce?
(Bug fix, feature, docs update, ...)

This addresses roughly 4 items in #440, relisted <details><summary> here</summary>

 1. fix the masking of the generic control variables. Currently they are all masked by maskC which is correct in most cases, but not in all
2. replace 3D masks by 2D masks where possible, e.g. use maskInC/W/S for 2D fields
3.  add @timothyas code for control variables for exchange coefficients in pkg/shelfice, this may include retiring ALLOW_SHIFWFLX_CONTROL
4.  add the ability to control the base 10 logarithm of a field

</details>

But I'll list them out slightly differently below. There's quite a bit going on, so I'm doing my best to make this outline something like a "review guide". I hope it is helpful.

1. Add shelfice transfer coefficients or iceshelf drag coefficient as control variables
    - if #SHI_ALLOW_GAMMAFRICT is not defined, `xx_shicoefft` and `xx_shicoeffs` are the relevant control parameters, which add modifications to `shiTransCoeffT` and `shiTransCoeffS`
    - if #SHI_ALLOW_GAMMAFRICT is defined, then `xx_shicdrag` is the relevant control parameter, and one can further decide to control only the thermodynamic impact of the drag coefficient, or both momentum and thermodynamic
    - I've included a check in `ctrl_check.F` to make sure the options above are specified correctly
    - This catches the case of a user adjusting only `xx_shicoefft` or only `xx_shicoeffs`, and appropriately adjusting the other based on the pkg/shelfice parameter `SHELFICEsaltToHeatRatio`, newly added in this PR to SHELFICE.h
    - <details><summary>The relevant files changed to simply enable the control variable are here:</summary>

        * ctrl_map_ini_genarr.F, specifically in the first subroutine CTRL_MAP_INI_GENARR
        * ctrl_init.F
        * ctrl_check.F
        * shelfice_init_varia.F
        * shelfice_readparms.F
        * shelfice_u/v_drag_coeff.F
        * src/the_main_loop.F <- this just adds some store statements to please TAF
        * src/initialize_varia.F <- also, just adds store statements
        </details>
    - <details><summary>These are the files necessary to make these controls work with the optimization, pack/unpack routines, these files also have some differences to :</summary>

         * ctrl_pack.F
         * ctrl_unpack.F
        </details>
    - <details><summary>These are the files changed to print diagnostics of the relevant adjoint variables added:</summary>

        * shelfice_init_fixed.F
        * pkg/autodiff/addummy_in_stepping.F
        * pkg/autodiff/adcommon.h
        * pkg/autodiff/common.flow
        </details>
    - I added `xx_shicoefft` verification/isomip/input_ad/data.ctrl
    - In these changes I got rid of the ALLOW_SHIFWFLX_CONTROL flags in a lot of places, but not all. In general, they are kept in the ECCO_DEPRECATED, and I didn't touch them in e.g. shelfice_thermodynamics.

2. Generically handle masks other than maskC for control variables, and the smooth package. This is closely linked to 1 because the shelfice parameters require maskCtrlI (created as maskSHI in shelfice_init_fixed), not maskC.
    - A note here is that in #406 `xx_gen*_grid` option was added as a user defined character to specify where on the grid a variable is, e.g. with `xx_gentim2d_grid = 'i'` for `xx_shifwflx`. My PR gets rid of this, and handles all of this under the hood. My reason is that there are already a lot of options for the user to keep track of, and I don't think this should be an *option* since there is technically one correct position for each control variable, so we might as well code it up and leave it. This gets rid of some stuff in CTRL_GENARR.h, and see changes to ctrl_init, ctrl_pack, ctrl_unpack, and grdchk_init. I'm curious what folks think of this (particularly @mjlosch who worked hard for #406).
    - The above point is separate from the masking issue, though, because e.g. in ctrl_map_ini_gentim2d and ctrl_map_ini_genarr2/3d, a mask is used to apply weights to the control variable. This is accomplished with the new routines ctrl_get_mask2d/3d in ctrl_get_mask.F
    - The ctrl_get_mask routines use some copy routines in the newly added ctrl_toolbox.F, which is a copy of some routines in the ecco_toolbox, except that this ctrl version fills overlaps.
    - <details><summary>The "generic masking" affects a lot of files, basically all of the control variable mapping (ctrl_map_*), calls to the smooth package, and where ctrl costs are computed. Here's the full list</summary>

        * ctrl_bound.F
        * ctrl_cost_driver.F
        * ctrl_cost_gen
        * ctrl_get_gen
        * ctrl_map_ini_gen (deprecated?)
        * ctrl_map_ini_genarr
        * ctrl_map_ini_gentim2d
        * pkg/ecco/cost_gencost_bpv4.F
        * pkg/ecco/cost_gencost_sshv4.F
        * pkg/ecco/cost_gencost_sstv4.F
        * pkg/ecco/cost_generict.F
        * pkg/smooth/smooth2d.F
        * pkg/smooth/smooth_basic2d.F
        * pkg/smooth/smooth_correl2d.F
        * pkg/smooth/smooth_diff2d.F
        </details>
    - verification/global_ocean.cs32x15 specifies `xx_fu` and `xx_fv` as having grid location as `w` and `s`, so I incorporated this into ctrl_init.F . However, the masks used for applying weights should be consistent, so this should probably be updated in `ctrl_get_mask2d`. I didn't do that yet because it opens a bigger can of worms: consider uvel/vvel as control variables, these are initialized in ctrl_init with `maskCtrlC`, but then they are masked with `maskW` and `maskS` ... I'm sure there are many cases of this inconsistency. 
3. This adds the ability to compute iceshelf melt rate or heat flux as objective functions in the generic cost framework of the ecco package, with just some minor updates to ecco_phys.F. Note that I also included a lot of comments to ecco_phys. I think they are helpful, but I'm totally fine with deleting them.
4. Allow a user to control the base 10 logarithm of a variable, rather than just the variable. Consider the example of a parameter which we guess to be `var = 10**-4`, but we'd like to adjust that parameter, we would be adjusting: `var = 10**( -4+ xx_var)` rather than the usual way, which would be: `var = xx_var + 10**-4`
    - A user specifies the option with `xx_genarr*d_c = 'log10ctrl'` 
    - Because control variables are initialized with zeros when `doInitXX=.TRUE.`, I thought a reasonable way to initialize these parameters differently is with `xx_genarr*d_r` option. For instance, in the example above one would set `xx_genarr2d_r = -4.`
    - There is an example of this in conjunction with `xx_shicoefft` in verification/isomip/input_ad/data.ctrl
    - Some changes to `ctrl_check` are made to ensure the user does not make conflicting choices, that is 'xx_gen*_preproc='noscaling'` conflicts with this option
     - <details><summary>The file changes for this are pretty small but listed here:</summary>

        * ctrl_check.F
        * ctrl_map_ini_genarr.F, specifically in the 2D and 3D mapping subroutines (a search with log10 will show the relevant lines)
        </details>
    - this could be separate from the rest of this PR, but it seems like not too much trouble to keep it in. If folks really want it to be separate, just let me know.

Documentation will need to be added, and I'm happy to do that once the general procedure is agreed upon.

## What is the current behaviour? 
(You can also link to an open issue here)

`xx_shifwflx` with gentim2d is not masked correctly so sensitivities and applied weights are inconsistent


## What is the new behaviour 
(if this is a feature change)?

This masking issue is fixed, along with the additional features listed above.

## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)

It doesn't seem like it. I have run adjoint tests isomip and global_ocean.cs32x15, and all forward tests. Testreport shows no difference to master. Additionally I compared the sensitivities to master, which you can see a comparison of [here](https://gist.github.com/timothyas/cee9946294ef0be202600c0f543b15fe)

## Other information:

A big shoutout to @gaelforget - this builds on his generic cost and ctrl infrastructure which I'm super thankful for 🙏

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

- add shiTransCoeffT, shiTransCoeffS, and shiCDrag as control variables
- generalize masking procedure for control parameters, fixes procedure for iceshelf controls
- add ability to compute shelfice freshwater and heat fluxes through ecco package
- add ability to control base 10 logarithm in generic control arrays

timothyas commented:
@ifenty and @owang01 - I know ECCO has been estimating the shelfice transfer coefficients, which I'm adding code for here (item 1 on the list above). I'd be curious what you think about the implementation, note that this doesn't just implement the control variables, but works with the pack/unpack routines for optimization too. I know it's a lot so I'd be happy to chat about it if you are interested.

mjlosch commented:
@timothyas I'll merge with master before I start reviewing. There have been some changes that lead to conflicts.

mjlosch commented:
@timothyas before I push my merged version, can you confirm that global_ocean.90x40x15.bottomdrag works for you? with 
```
testreport -devel -adm -ncad -t global_ocean.90x40x15
```
this experiment fails because of a floating point exception (`bottomdragfld` is sometimes Infinity). I am not sure if this is an optimisation problem or not, because I can make this error go away when I print `bottomdragfld` after filling it in `ctrl_map_ini_genarr`.

timothyas commented:
Thanks for looking into this @mjlosch! I ran the testreport you posted with this branch and it seems to be fine:

```
Y Y Y Y 16>13< 9 pass  global_ocean.90x40x15  (e=0, w=9)
Y Y Y Y 16>14<10 pass  global_ocean.90x40x15.bottomdrag
Y Y Y Y 16>14< 1 pass  global_ocean.90x40x15.kapgm
Y Y Y Y 16>14< 3 pass  global_ocean.90x40x15.kapredi
```

Which is no different from a re-run with an old master branch.

mjlosch commented:
@timothyas OK, thanks, then something got really screwed up during the merge with the master branch ... I'll start again.

mjlosch commented on pkg/ctrl/ctrl_get_gen.F:
```diff
@@ -196,7 +201,7 @@ subroutine ctrl_get_gen(
 #ifdef ALLOW_SMOOTH
 #ifdef ALLOW_SMOOTH_CTRL2D
-        if (useSMOOTH) call smooth2D(xx_gen1,genmask,1,mythid)
+        if (useSMOOTH) call smooth2D(xx_gen1,mask2D,1,mythid)
```
what keeps us from replacing `genmask` entirely by `mask2D` also in the DEPRECATED code?

timothyas replied:
Do you mean, lets get rid of the `genmask` argument and just use this routine to get the mask within `ctrl_get_gen`? If so, I agree.

It seems like this routine is only called from `ctrl_map_gentim2d` where `genmask` always takes the value of `maskC`. This would replace that.

mjlosch replied:
Yes, genmask is the only 3D array in this routine, so we can either replace it by a 2D mask (and then call ctrl_get_gen with a 2D mask) or get rid of it entirely and get the mask from within `ctrl_get_gen` (your changes make `genmask` entirely superfluous already).  I am not sure what I like better, but I also tend towards your solution of getting the mask from within the routine. 
In any case, there would be additional changes to `exf_getclim.F`, `exf_getffields.F`, `exf_getsurfacefluxes.F`, and `shelfice_thermodynamics.F` , as far as I can see.

timothyas replied:
Done

mjlosch commented on pkg/ecco/cost_gencost_bpv4.F:
```diff
@@ -115,6 +116,9 @@ subroutine cost_gencost_bpv4(
              locbpmask(i,j,bi,bj) = 0. _d 0
+
+             ! For smoothing routines
+             mask2D(i,j,bi,bj) = maskC(i,j,1,bi,bj)
```
shouldn't this be `maskInC` instead of `maskC(:,:,1,:,:)`?

timothyas replied:
I'm not sure of the difference actually, but yep that makes sense. The old code had `maskC(:,:,1 ..)` so I was just copying that.

mjlosch commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -441,6 +442,8 @@ subroutine cost_gencost_sshv4(
           do j = jmin,jmax
             do i = imin,imax
+               ! mask2D for smoothing. here to avoid TAF recomp
+               mask2D(i,j,bi,bj) = maskC(i,j,1,bi,bj)
```
same as before: Isn't `maskInC` the better choice?

mjlosch commented on pkg/ecco/cost_gencost_sstv4.F:
```diff
@@ -173,6 +174,9 @@ subroutine cost_gencost_sstv4(
               msk_sst(i,j,bi,bj)  = 0. _d 0
+
+              ! mask2D for smoothing
+              mask2D(i,j,bi,bj) = maskC(i,j,1,bi,bj)
```
`maskInC`?

mjlosch commented on pkg/shelfice/shelfice_init_fixed.F:
```diff
@@ -279,6 +286,31 @@ SUBROUTINE SHELFICE_INIT_FIXED( myThid )
 #endif
 
+# ifdef ALLOW_AUTODIFF
+#  ifndef SHI_ALLOW_GAMMAFRICT
```
This will throw an error for the wrong combination of CPP-flags and entries in data.diagnostics. Is that intended? Alternatively one could define all three diagnostics and then some are filled with zeros if specified with the wrong CPP flag.

timothyas replied:
I do prefer to throw an error rather than passively fill zeros, so the user doesn't get the false satisfaction of a completed run all to find the field is just zero... That said the error raised here might be cryptic, and I could add a check in `shelfice_check.F`

timothyas replied:
Ah I see that the forward model does the behavior you're mentioning: passive failure by filling 0's rather than raise an error. However, I would be in favor of putting both the forward and adjoint within the CPP macros and to raise an error rather than fill with 0's. This seems to be the behavior in other packages, e.g. in gmred_diagnostics_init, mom_diagnostics_init

mjlosch commented on pkg/smooth/smooth_check.F:
```diff
@@ -0,0 +1,80 @@
+C --- make sure proper macros are defined for maskI (iceshelf domain)
+        IF (smooth2DmaskName(smoothOpNb)(1:5).eq.'maskI') then
+          WRITE(msgBuf,'(A,A,I2,A,A,A,A)') 'SMOOTH_CHECK: ',
+     &      'Must compile with shelfice package, ALLOW_SHELFICE'
```
There is a `PRINT_ERROR( msgBuf, myThid )`  missing

mjlosch commented on pkg/smooth/smooth_correl2dw.F:
```diff
@@ -31,7 +31,7 @@ subroutine smooth_correl2Dw (
 #endif
 
-      _RL mask_in(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nR,nSx,nSy)
+      _RS mask_in(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nR,nSx,nSy)
```
why is this still 3D?

timothyas replied:
I actually could not find an instance of this subroutine being used, so I left it. But I agree, might as well change it while we're here.

mjlosch commented on pkg/smooth/smooth_diff2d.F:
```diff
@@ -38,11 +37,11 @@ subroutine smooth_diff2D (
 
-c for now: useless, because level 3 is recomputed anyway 
+c for now: useless, because level 3 is recomputed anyway
 c but : if level3 was computed during the fwd loop by callung
```
spelling -> calling

mjlosch commented on pkg/smooth/smooth_filtervar2d.F:
```diff
@@ -1,4 +1,11 @@
 #include "SMOOTH_OPTIONS.h"
+#ifdef ALLOW_CTRL
+# include "CTRL_OPTIONS.h"
```
Do we need this here?

timothyas replied:
It's definitely not, that's why I wrote the smooth2dmask = ... if/else branch at the top of this routine. Thanks!

mjlosch commented on pkg/smooth/smooth_filtervar3d.F:
```diff
@@ -1,4 +1,7 @@
 #include "SMOOTH_OPTIONS.h"
+#ifdef ALLOW_CTRL
+# include "CTRL_OPTIONS.h"
```
probably not needed?

timothyas replied:
Same, it's not - good catch.

timothyas commented:
@mjlosch I can also give it a shot if you like, just say the word

mjlosch commented:
@timothyas 
> I can also give it a shot if you like, just say the word
Please don't, I am almost done and will push soon (and I tried to make sure that there are plenty of @jahn in the commit messages (o: )

jahn commented:
Thanks, I appreciate that a lot.  You can never get enough email!

mjlosch reviewed (commented):
In general, this looks good to me. There are a few small changes that I will push in a few moments and there a comments that we can address or not, depending on how important we think they are.

mjlosch commented on pkg/autodiff/addummy_in_stepping.F:
```diff
@@ -457,6 +496,17 @@ SUBROUTINE ADDUMMY_IN_STEPPING( myTime, myIter, myThid )
           CALL MNC_CW_RL_W('D','adstate',0,0,
      &                     'adkapredi', adkapredi, myThid)
+#endif
+#ifdef ALLOW_SHIELFICE
```
fixed

timothyas replied:
Thanks!

timothyas commented on pkg/ecco/ecco_phys.F:
```diff
@@ -220,22 +241,73 @@ subroutine ecco_phys( mythid )
-          enddo
+          enddo ! Ends do k=1,nr
 c---------
+#ifndef ALLOW_SHELFICE
```
You're correct, this is not necessary. I wrote this many moons ago, I'd suggest this change for the m_eta, m_shelfice block

```fortran
          tmpmsk  = 0. _d 0
          tmpfld  = 0. _d 0
          tmpmsk2 = 0. _d 0
          if (gencost_barfile(kgen)(1:13).EQ.'m_boxmean_eta') then
            tmpmsk=maskC(i,j,1,bi,bj)*gencost_mskCsurf(i,j,bi,bj,kgen)
            tmpfld=m_eta(i,j,bi,bj)
            if (tmpmsk.NE.0. _d 0) tmpmsk2=1. _d 0
          endif
#ifdef ALLOW_SHELFICE
          if((gencost_barfile(kgen)(1:16).EQ.'m_boxmean_shifwf').or.
     &       (gencost_barfile(kgen)(1:16).EQ.'m_boxmean_shihtf')) then

            tmpmsk=maskSHI(i,j,1,bi,bj)*
     &             gencost_mskCsurf(i,j,bi,bj,kgen)

            if (gencost_barfile(kgen)(11:16).EQ.'shifwf') then
              tmpfld=shelficeFreshWaterFlux(i,j,bi,bj) / rhoConstFresh
            elseif (gencost_barfile(kgen)(11:16).EQ.'shihtf') then
              tmpfld=shelficeHeatFlux(i,j,bi,bj)
            endif
            if (tmpmsk.NE.0. _d 0) tmpmsk2=1. _d 0
          endif
#endif /* ALLOW_SHELFICE */
```

I can push it if you agree.

mjlosch replied:
please go ahead!

timothyas replied:
Done

timothyas commented on pkg/ctrl/ctrl_map_ini_gen.F:
```diff
@@ -316,6 +317,7 @@ subroutine ctrl_map_ini_gen2D(xxFileCur, wFileCur, xxDummyCur,
          DO i = 1-OLx,sNx+OLx
            xxFld2d(i,j,bi,bj)=0. _d 0
            wFld2d(i,j,bi,bj)=0. _d 0
+           mask2D(i,j,bi,bj)=maskFld3D(i,j,1,bi,bj)
```
Good call, we could pass maskInC where this is called in `ctrl_map_ini_ecco`. I did not pay close attention to this because I believe it is deprecated.

timothyas replied:
this seems to be a repeat, resolving one

timothyas commented on pkg/ctrl/CTRL_OPTIONS.h:
```diff
@@ -66,6 +66,7 @@ C       >>> Other Control.
 #undef ALLOW_KAPGM_CONTROL
 #undef ALLOW_KAPREDI_CONTROL
 #undef ALLOW_BOTTOMDRAG_CONTROL
+#undef ALLOW_SHIFWFLX_CONTROL
```
I added it at some point just to have a more complete list, but I agree lets just get rid of it

timothyas commented:
Thanks for the in-depth review @mjlosch. I agree with all of the questions/suggestions you raised and implemented most fixes - can I push those? 

Also I just thought I'd let you know I did some diff'ing just to make sure the merge master went ok and it all looks good - thanks for your steady hand there

timothyas commented on pkg/ctrl/ctrl_get_mask.F:
```diff
@@ -0,0 +1,149 @@
+      if ((xx_filename(1:11).EQ.'xx_shicoeff').or.
+     &    (xx_filename(1:11).EQ.'xx_shicdrag').or.
+     &    (xx_filename(1:11).EQ.'xx_shifwflx')) then
+        call ctrl_cprsrs(maskSHI,Nr,mask3D,Nr,myThid)
```
I don't think we can get rid of `maskSHI`, or at least some field which denotes the 2D iceshelf extent. This will be essentially the complement of `maskInC` in a domain with an iceshelf.

As for `ALLOW_SHIFWFLX_CONTROL`, the only block of code that seems problematic is this one:

https://github.com/MITgcm/MITgcm/blob/c512e371cc44c0dc594af9f2af15cc8a957c15a0/pkg/shelfice/shelfice_thermodynamics.F#L249-L260

but this is deprecated, so perhaps we want to get rid of it anyway? Otherwise, maybe we can otherwise add a block of code in `shelfice_init_fixed.F`, e.g. 

```fortran
doSHIFwflxCtrl = .FALSE.
#ifdef ALLOW_GENTIM2D_CONTROL
DO k2 = 1, MaxCtrlTim2d
    IF xx_gentim2d_file(k2)(1:11).EQ.'xx_shifwflx') doSHIFwflxCtrl = .TRUE.
ENDDO
#endif
```

with boolean `doSHIFwflxCtrl` defined in `SHELFICE.h`. Then in shelfice_thermodynamics

```fortran
#ifdef ALLOW_GENTIM2D_CONTROL
IF (doSHIFwflxCtrl)
... do something ...
ENDIF
```
#endif

What do you think?

mjlosch replied:
what I meant was this: get rid of the 3D mask and use 2D mask instead. I think we introduced it initially as a hack (as the comments in the code imply), but it just looks weird to me to have a 3D mask that essentially masks 2D arrays. Your suggestion seems to go into this direction, right? Maybe something to discuss in a call?

timothyas replied:
Ah, I see make `maskSHI` 2D. As you mention in the comments though, this might be tricky for the pack/unpack routines. Or at least more would have to change. I think something to discuss in a call, I'll check out what this would require.

mjlosch replied:
let's defer this to later PR

mjlosch commented:
@timothyas 
please go ahead and push your changes. I'll re-review.

timothyas commented:
I have no idea why the openad test is failing and it is extremely cryptic to debug but I do not have the time for it now. Will try later.

Otherwise, most other changes from @mjlosch 's review have been implemented.

mjlosch commented on pkg/ctrl/ctrl_get_gen.F:
```diff
@@ -76,9 +75,11 @@ subroutine ctrl_get_gen(
 #ifdef ALLOW_SMOOTH
 #ifdef ALLOW_SMOOTH_CTRL2D
       _RS dummyRS(1)
+
```
blank line intentional?

mjlosch commented on pkg/ctrl/ctrl_smooth.F:
```diff
@@ -14,7 +14,7 @@ SUBROUTINE CTRL_SMOOTH (
 c     input/output
 c     fld    : 2-D array to be smoothed
       _RL fld ( 1-OLx:sNx+OLx,1-OLy:sNy+OLy,   nSx,nSy )
```
one could align this now.

mjlosch commented on pkg/ecco/ecco_phys.F:
```diff
@@ -220,22 +241,68 @@ subroutine ecco_phys( mythid )
+          tmpfld  = 0. _d 0
+          tmpmsk2 = 0. _d 0
           if (gencost_barfile(kgen)(1:13).EQ.'m_boxmean_eta') then
+            tmpmsk=maskC(i,j,1,bi,bj)*gencost_mskCsurf(i,j,bi,bj,kgen)
```
shouldn't this be `maskInC`

mjlosch commented on pkg/smooth/smooth_correl2dw.F:
```diff
@@ -31,7 +31,7 @@ subroutine smooth_correl2Dw (
 
-      _RL mask_in(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nR,nSx,nSy)
+      _RS mask_in(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
       _RL fld_in(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
```
alignment?

mjlosch commented:
@timothyas this looks great. I have made a few comments (sorry for not doing them in one review). I can push related cosmetic changes, but maybe better wait until the OpenAD issue is fixed?  I am afraid I cannot help with that.

Let's talk about the unresolved conversations in a call, sounds easier.

timothyas commented:
Ok I did the check pass dance and it worked! I don't really understand why the change was needed, but I had to add `mask2D` as an argument to `CTRL_GET_GEN` and `CTRL_INI_GEN2D`. Actually I'm not sure if both were necessary, or if I just had a bug all along (although I doubt that because the TAF tests worked on my machine...), but I'm not going to look any further because this works.

As for the review, thanks for the comments. Are these the changes you want to push? I'll wait for you to push, and anything minor that isn't implemented I'll go ahead and clean up. Then, agreed, we can chat about the more complex stuff.

Thanks @mjlosch!

mjlosch commented:
@timothyas 
I fixed a file that didn't compile, and I change ctrl_get_mask to avoid extra copy operations, maybe we can do something similar for the 2D case (I apologize for being incomplete here). The rest are the cosmetic changes from my review.

timothyas commented:
No problem, done

timothyas commented:
@mjlosch stray comments have been removed, and it turns out the other OpenAD problems were that there was an `#ifdef ALLOW_OPENAD` blocking the definition of `k2` and `iarr` in `ctrl_check` - this caused the earlier check failures. Now I am simply defining these variables only if generic controls are used.

It seems like the generic controls are not guaranteed to work with OpenAD - there are many cases where generic control operations are only defined if `ALLOW_OPENAD` is not defined. I think this would take a much more focused effort to make consistent, and it would have to be done by someone who uses OpenAD, i.e. not me :)

mjlosch commented:
@timothyas 
sorry for these late running comments: 
1. I think that we could rename `mask2D` in `ctrl_get_gen` (back) to `genmask`, then the changes to this file would be minimal (compared to upstream/master).
2. would it make sense to replace 
 ```CALL CTRL_CPRSRS(maskC, Nr, mask3D, Nr, myThid)``` 
 with
  ```CALL CTRL_CPRSRS(maskC, Nr, mask2D, 1, myThid)``` 
  for the 2D case in ```SUBROUTINE CTRL_GET_MASK2D```? This way we would avoid copying 3D fields when only 2D fields are required.

If you agree with these suggestions, I can make these changes.

timothyas commented:
No problem @mjlosch, those are both good suggestions - thanks!

timothyas commented:
@jm-c and @mjlosch: I've added documentation for the added capabilities, this should be good to go.

mjlosch commented on doc/ocean_state_est/ocean_state_est.rst:
```diff
@@ -868,6 +889,21 @@ to the model grid unless CPP-flag :varlink:`EXCLUDE_CTRL_PACK` is defined in
   +-----------------------+-----------------------+-----------------------+
 
+
+.. table:: ``xx_gen????d_preproc_c`` options implemented as of checkpoint
```
I guess you put this here to give some incentive to also document the other options? 👍

timothyas replied:
I actually struggled with rst formatting for a while trying to add a separate section to the existing `xx_gen*_preproc` table, but could not get the table to work out. After a while I gave up and made a new one...

timothyas replied:
I added a section about the log10 controls just to be as explicit as possible, which should be the last "content" update.

mjlosch replied:
My initial comment was totally ignorant. I didn't realise that there was already a table with preprocessing options. There was no irony or sarcasm intend in this comment, sorry if it sounded like that.

I pushed a small change so that the two tables are merged. See, if that's OK.

Otherwise all of this looks good to me.

timothyas replied:
No problem @mjlosch! I didn't read it as sarcastic at all. There's always more documentation to be written :)

I would probably prefer it the other way, though. It seems somewhat unclear to me how it currently is written because the table is for `xx_gen*_preproc` options, while the log control is a `xx_gen*_preproc_c` option, i.e. an option based on characters (along with the `_i` and `_r` integer and real options). Right now there are no other `_c` options so the table looks a little funny, but I was hoping to be totally clear on that one. Basically I was having trouble making two "sub-tables" within the one `preproc` table, with different sub-sections for `_c` (etc) options. I was just unable to merge rst table cells together for "sub-headings".

What do you think?

mjlosch replied:
I think I don't quite understand what you intended to do with this table, please feel free to revert my change.

Is this the first `_c` option that is introduce to this? I couldn't find anything else. In any case, this is nothing to get hung up about, and I am OK with it either way.

timothyas replied:
It is the first `_c` option, so while the table looks a little funny on its own, I think it's appropriate. I'll go ahead and revert.

timothyas replied:
Reverted via force-push

timothyas replied:
With @jrscott's doc approval, I'll go ahead and resolve this

mjlosch commented on doc/ocean_state_est/ocean_state_est.rst:
```diff
@@ -911,6 +947,38 @@ The user must be mindful of control parameter combinations that make sense
+coefficient used in the simulation.
+
+The adjustments ``xx_shicoefft`` and ``xx_shicoeffs`` are available with the
+velocity **independent** form of transfer coefficients is used, by setting
```
there's something wrong I'll come up with a suggestion

timothyas replied:
Thanks. I see your commit. I didn't know about the 80 character limit. For some reason my terminal+tmux+vim freaks out with rst so have been using a notepad text editor with no auto line adjustor

mjlosch replied:
There is none, except that Oliver once cited a recommendation in a different context. It's an unsafe subject :)

mjlosch commented on doc/phys_pkgs/shelfice.rst:
```diff
@@ -107,9 +110,12 @@ that is different from :math:`\rho_{0}`, the anomaly is zero.
    | :varlink:`SHELFICEheatTransCoeff`      | SHELFICE_PARM01 | 1.0E-04                                    | transfer coefficient (exchange velocity) for temperature (m/s)                                          |
    +----------------------------------------+-----------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------+
-   | :varlink:`SHELFICEsaltTransCoeff`      | SHELFICE_PARM01 | 5.05E-03 :math:`\times`                    | transfer coefficient (exchange velocity) for salinity (m/s)                                             |
+   | :varlink:`SHELFICEsaltTransCoeff`      | SHELFICE_PARM01 | :varlink:`SHELFICEsaltToHeatRatio`         | transfer coefficient (exchange velocity) for salinity (m/s)                                             |
```
this leads to the table running over the margins. I don't know how to fix it, maybe by splitting the parameters across lines? But then the `varlink` doesn't work anymore.

mjlosch replied:
in my build the last column extends much further into the "grey zone", but I agree with you, there's not much one can doe.

timothyas replied:
I looked at the current online version and the "grey zone" is already there, so I wasn't too worried about it. If anyone has any specific guidance here then I'm more than happy to make a change, but otherwise I'll leave it.

mjlosch replied:
Yes, I agree, but maybe @jrscott has an idea?

jrscott replied:
The table boxes all need to be the same size. You can break text into multiple lines, but it often easier just to have the tables go much longer lines than 80 char, which is what is typically done.

In any event, I'll fix it all up when I review the doc changes, I unfortunately have a lot of experience fighting rst to make some look sensible. Just drop a note that the doc is ready for review.

jrscott replied:
So I think the problem here is that there is simply too many letters (in a row) for the table to display (when long variable names are in columns), and it handles it is a weird way -- instead of making all rows a bit longer it seems to be making every other row wider. I gave it a go, was not able to fix properly :(

Suggest: since all the run-time options are namelist SHELFICE_PARM01 (and there isn't even a  SHELFICE_PARM02 at this time) can we eliminate this column? I'll specify namelist SHELFICE_PARM01 in the table heading line.

timothyas replied:
That sounds good to me - thanks @jrscott!

mjlosch reviewed (approved):
Looks good to me. If you have fix for the table in section "8.6.3.3. SHELFICE run-time parameters", let's have it, but if not, I don't know a fix to this, either.

ifenty commented:
@timothyas all of this work on shelf is amazing. Is it correct that now that shelf is main package works in ad mode? Because it did not when I tried and failed -- @mjlosch -- to merge our new combined ice shelf/ice front code. 

To see this development work happening is a joy and also incredibly frustrating because I believe what we tried to merge also included control parameters on the shelf ice heat flux coefficients. I can say at least that the adjointable shelf ice/ice front code that we're using now for at ECCO V4r5 does have a heat flux coefficient control parameter. The salt flux coefficient is a function of the heat flux coefficient if I'm not mistaken.

Since it seems you are in deep with the shelf ice code now, would you be amenable to a conversation to see if we can chart a path forward on integrating the innovations from our combined shelf ice/ice front code and getting that PR accepted?

I know one issue is the formulation of the three equation solver, @mjlosch pointed out that in its current formulation it will not work in a vector machine or not be efficient in the vector machine, that might be a simple fix I'm not exactly sure what needs to be done though.  There is another issue with where the heat flux tendencies and salt flux tendencies are applied, with the current shelf ice code there are calls to surface forcing T and surface forcing S, well in the new version I operate on gT and gS directly everywhere and make use of addmass everywhere.

mjlosch commented:
@ifenty 
one big difference between this PR and PR #368 is that #368 fails the regression test(s).
BTW, the shelfice package did work with the adjoint all the time (there are AD-regression tests with isomip), just not with the generic control variables, which is what @timothyas fixed (among other things) in the PR. But I acknowledge that there are still specific things where the adjoint is stopped (e.g. for `SHI_useRealFWflux=T` in shelfice_forcing_surf.F ). I don't know why this was done this way, but why not try to lift that restriction in the available code rather than writing a new package?
Only because you mentioned the vectorisation: The changes in this PR are not the least as radical as in #368 and there are many other concerns with that PR that are listed in #368 and even mores in #378. The vectorisation is probably the smallest problem and can most likely be fixed without too much effort.

timothyas commented:
@ifenty regarding shelfice not working with the adjoint, I noticed this as well in #349, specifically see 1. in my comment [here](https://github.com/MITgcm/MITgcm/issues/349#issue-625196399). I see no reason for the STOP, but this can be addressed easily in a separate PR. That is, unless anyone knows of a reason why it is there, and given @mjlosch has mentioned he does not, I will venture a guess to say that it doesn't need to be there.

ifenty commented:
> @ifenty regarding shelfice not working with the adjoint, I noticed this as well in #349, specifically see 1. in my comment [here](https://github.com/MITgcm/MITgcm/issues/349#issue-625196399). I see no reason for the STOP, but this can be addressed easily in a separate PR. That is, unless anyone knows of a reason why it is there, and given @mjlosch has mentioned he does not, I will venture a guess to say that it doesn't need to be there.

No, I don't know why the STOP is there.  Have you been using the adjoint of the shelf is model for your work? If so, does it behave as expected?

timothyas commented:
I have been using it @ifenty. It seems to make sense to me. I am not using the particular switches for which the "STOP" statement matters though

mjlosch commented:
@timothyas 
can you produce the TAF error mentioned in #446 with any of the verification experiments? I guess not, because none of them uses it.

mjlosch commented:
@timothyas
I guess the problem is a little more subtle. There are places in pkg/smooth, where `smooth2Dmask` is set according to `smooth2DmaskName` in "smooth_filtervar2d.F", but also only within `j = 1,sNy` and `i = 1,sNx`, just like in the old version of "smooth_diff2d.F", where `smooth2Dmask` was a copy of `mask_in` with `k=1`, but only within `j = 1,sNy` and `i = 1,sNx`, so the exchange was necessary. Now, as we pass the 2D mask directly we just need to make sure that is does not need an exchange and remove the problematic line. 
I have found these places where `smooth_diff2d` is called:
```
> grep -i "smooth_diff2d" *.F | grep -i call
smooth2d.F:      call smooth_diff2D(fld_in,mask_in,nbt_in,mythid)
smooth_basic2d.F:C     |   calls smooth_diff2D.F.
smooth_basic2d.F:      call smooth_diff2D(fld_in,mask_in,nbt_in,mythid)
smooth_correl2d.F:      call smooth_diff2D(fld_in,mask_in,nbt_in,mythid)
smooth_filtervar2d.F:c      call adsmooth_diff2D(smoothTmpFld,smooth2dmask,nbt_in,mythid)
smooth_filtervar2d.F:      CALL smooth_diff2D(smoothTmpFld,smooth2dmask,nbt_in,mythid)
smooth_filtervar2d.F:      call smooth_diff2D(smoothTmpFld,smooth2dmask,nbt_in,mythid)
smooth_hetero2d.F:C     |   calls smooth_diff2D.F.
smooth_hetero2d.F:      call smooth_diff2D(fld_in,mask_in,nbt_in,mythid)
```
- `smooth2d` is called from `ctrl_get_gen.F`, `ctrl_map_ini_gen*.F`, all of which have been checked in this PR, so that should be safe.
- `smooth_basic2d` is called from `cost_gencost_bpv4.F` with `maskInC`, so that's safe.
- `smooth_correl2d` is called from `ctrl_map_ini_*.F` which have been modified in this PR, so safe (all called from ctrl_map_ini_ecco).
- `smooth_filtervar2d` needs to be modified (see above)
- `smooth_hetero2d` is called from `cost_gencost_ss*v4.F` with `maskInC` and from `cost_generic.F` with `mask2D`, which is computed over the entire domain from `maskC/S/W`, so also safe.

So all we need to do is fix smooth_filtervar2d.F and remove the exchange in smooth_diff2d.F
What do you think?

timothyas commented:
Hey @mjlosch, thanks for the detailed overlook. You are right, the overlaps need to be filled in filtervar. I modified both the 2d and 3d to do this. Otherwise, I think all routines are good as far as overlaps go.

The TAF error can be reproduced simply by adding `smooth` and `ecco` to the `isomip` verification experiment. I was able to produce the error, and then get rid of it by removing the mask exchange. I removed the exchange in a single commit just to isolate it, but I think it makes sense to get rid of that line.

While making these changes I found a small typo in `smooth_check` - now fixed. The smooth package should probably be tested... Perhaps a contribution I can make some day, maybe after graduation ...

jrscott commented:
@jrscott to review doc changes

jrscott commented:
fixed up tables in both html and latex. doc LGTM

timothyas commented:
Thanks @jrscott!

I'll merge master into this branch to resolve conflicts with `shelfice_init_varia` likely from merging of #465. I'll run some more tests as mentioned and ping folks when that is done.

timothyas commented:
@mjlosch @jm-c I feel quite confident about this PR now. This passes all adjoint tests on my linux machine to the same precision as upstream/master. Additionally, I tested verification_other/global_oce_cs32 which uses ecco, exf, ctrl, and smooth together. I had to make one small change from that in `cost_generic`, specifically `S/R cost_genloop` to create `localmask` and `mask2D` with their overlaps filled (note that `ecco_cprsrl` in `ecco_toolbox` does not fill overlaps, I made a comment with this note in the subroutine, stating why the masks were being filled there explicitly).

timothyas commented:
@mjlosch @jm-c I have merged master into this branch, which includes Martin's epic TAF Recomputation PR. I'm happy to report that this branch continues to pass all adjoint tests to the same degree as upstream/master, and also, a more fine grained test between this branch and master shows:
- no differences in forward_step_ad.f other than some common blocks in the isomip test
- no differences in taf_ad.log other than some extra files to parse
- the `mitgcmuv_ad` binaries are about the same size

Additionally, I've gotten rid of the recomputations that used to exist when the velocity dependent code is compiled, i.e. with `#define SHI_ALLOW_GAMMAFRICT` (see 3bb49dbe81).

mjlosch commented:
I had a quick look, and can confirm that this looks good!

Nice (cheap in terms of memory!) hack to convince TAF into not doing recomputations.

jm-c commented:
@timothyas I made little minor changes. Please take a look and could adjust these if needed.
Otherwise, all changes look good. One minor comment: new S/RCTRL_ZERO (in new file "ctrl_toolbox.F") could be renamed
since it reset the field to a constant value which is not necessary zero as it is passed as an argument.

timothyas commented:
Hey @jm-c thanks for your updates, it looks good to me. I see some possible entries for the highly anticipated "MITgcm Style Guide" ...🙂

I renamed `CTRL_ZERO` -> `CTRL_ASSIGN` in the latest commit - I agree it should be renamed.

timothyas commented:
The latest commit with updates to `adcommon.h` does not affect results, it just corrects the diagnosed sensitivities in `addummy_in_stepping.f`. The updates in #465 changed TAF's adjoint shelfice common blocks, and this updates the transformation of `var_ad` -> `advar`, which is necessary for `addummy_in_stepping.f` because it calls `ad` prefixed variables.

We should probably just eventually call variables by the suffix `_ad` as TAF generates them, then `adcommon.h` wouldn't be necessary. Ah, the never ending AD todo list...

jm-c commented:
@timothyas I also forgot 1 file this morning and just push it. I also added the tag-index entry so, from my side, this is ready to be merged in. Could do that later today.

timothyas commented:
Thanks @jm-c. I'm not sure how those `_OPTIONS.h` CPP header files made it inside those subroutines. I did a search and found one more.

With that, I think it's good to go. Thanks!

jm-c commented:
I think there is a problem with field `shiDragQuadFld` C-grid location: it's used as if it was at grid-cell western-edge (uVel position) in `shelfice_u_drag_coeff.F` like here:
https://github.com/MITgcm/MITgcm/blob/51e381e9c9a1bef17c6feb83a95af189631764de/pkg/shelfice/shelfice_u_drag_coeff.F#L141
but also at the grid-cell southern-edge (vVel position) in `shelfice_v_drag_coeff.F` and also filled up directly from a grid-cell centered field `shiCDragFld` in `pkg/ctrl`.

It does not matter when the drag-coeff is not used as a control (just set to a constant in this case)  but otherwise it could have some unexpected consequences (e.g., losing water with CS-grid and ice-shelf located across 2 faces).
